---
title: "PANTS Vignette"
author: "Jonathan Dreyfuss"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction
Welcome to the Pathway analysis via network smoothing (`PANTS`) R package. In this vignette, I'll illustrate a case study of the package for integrating proteomics and metabolomics data to identify significant pathways.

## Install & load
See README.md.

```{r lib, message=FALSE}
library(limma)
library(ezlimma)
library(PANTS)
```

## Load interaction network
We'll use the Pathway Commons v9 network in Simple Interaction Format (SIF), which holds interactions in the format: physical_entity_id <relationship type> physical_entity_id. The 1st and 3rd columns represent the edges of the network, an edgelist. This edgelist has been included in `PANTS`, so you can load it using:
```{r pc9}
data("pc9")
```

## Construct graph
`PANTS` smooths statistics over the network using a network kernel derived from a graph. So we transform `pc9` into a simplified graph:
```{r gr}
gr <- igraph::simplify(igraph::graph_from_edgelist(pc9, directed = FALSE))
```

## Construct predefined pathways
We need to load predefined pathways that consist of genes and metabolites whose IDs are consistent with those of our interaction network, and transform these into a matrix that `PANTS` accepts. One source for these pathways is the Small Molecule Pathway Database (www.smpdb.ca). SMPDB offers one file for proteins and another for metabolites. I've alread read these and constructed pathways and included the result with the package. You can load this data with:
```{r gmat}
data("smpdb_gmat")
```

You read in an arbitrary pathway file in GMT format with `ezlimma::read_gmt` and transform it with `PANTS::gmt2Gmat`.

## Calculate Laplacian kernel matrix
This matrix is used for smoothing scores across the network, like how heat diffuses. It can be a dense matrix, a sparse matrix from package `Matrix`, or `NULL` to avoid smoothing.

```{r ker}
ker <- graph2kernel(gr)
```

## Get data
To illustrate `PANTS`, we'll simulate protein and metabolite abundance data on 10 samples, with 5 cases and 5 controls. The rownames (feature names) of this data matrix need to overlap with the rownames of `gmat`. We'll sample features from `gmat` instead of using all the features to save computation time. We also need a vector of phenotypes.

We treat this data as being already processed. For many datasets, there are zeroes or missing values that might need to be imputed; samples need to be normalized to make them comparable and amenable to statistical tests; absent analytes need to be removed; sample outliers need to be assessed to examine whether some experimental variables or batch effects need to be accounted for, or the samples need to be removed or down-weighted; and trends between an analyte's mean expression and its variance should be accounted for, especially in RNA-seq data, with `limma`'s `voom` function.

```{r sim}
set.seed(42)
features <- rownames(smpdb_gmat)[sample(x=1:nrow(smpdb_gmat), size=10^3)]
M <- matrix(rexp(n=10^4), nrow=10^3, ncol=10, dimnames=list(features, paste0("s", 1:10)))
pheno.v <- setNames(rep(c("case", "ctrl"), each=5), nm=colnames(M))
```

## Pathway analysis
We execute the main function, `pants`, by testing a contrast (or comparison) on this dataset, e.g. case vs. control. We limit the number of permutations to speed up the calculation. In general, if the `object` or `nperm` is large, this step will be slow.
```{r pants}
contr <- c(CASEvsCTRL="case-ctrl")
res.lst <- pants(object=M, phenotype=pheno.v, contrast.v=contr, ker=ker, Gmat=smpdb_gmat, nperm=50)
pwy.stats <- res.lst$pwy.stats
```

## Plot
Draw network diagram of driving nodes for the pathway "Beta-Alanine Metabolism". We annotate select nodes we have annotation for to replace their ChEBI ID with a name.
```{r plot}
zscore.v <- stats::setNames(res.lst$feature.stats$z, nm=rownames(res.lst$feature.stats))
annot <- c("CHEBI:15846"="NAD+")
plot_pwy(gr=gr, ker=ker, Gmat=smpdb_gmat, pwy="Beta-Alanine Metabolism", zscore.v=zscore.v, annot.v=annot, ntop=5, name=NA)
```